### PR TITLE
Ensure to reset the smoothing plugin when resuming Servo

### PR DESCRIPTION
**Pull request by sea-bass**
This fixes another Servo issue we found when resuming Servo after moving through another means, e.g., by running the motion planner to move to a completely different waypoint.

With this fix, the robot shouldn't teleport in these situations.

@ibrahiminfinite

### CODE REVIEW

File: moveit_ros/moveit_servo/src/servo_node.cpp [hunk 1]
```diff
@@ -162,6 +162,9 @@ void ServoNode::pauseServo(const std::shared_ptr<std_srvs::srv::SetBool::Request
   }
   else
   {
+    // Reset the smoother with the robot's current state in case the robot moved between pausing and unpausing.
+    servo_->resetSmoother(servo_->getCurrentRobotState());
```
Comment (sea-bass) [inline moveit_ros/moveit_servo/src/servo_node.cpp:166]: this is the actual fix, the rest is reducing code duplication

File: moveit_ros/moveit_servo/src/servo.cpp [hunk 2]
```diff
@@ -630,16 +640,12 @@ std::pair<bool, KinematicState> Servo::smoothHalt(const KinematicState& halt_sta
...
 
+  updateSmoother(target_state);
+
   // If all velocities are near zero, robot has decelerated to a stop.
   stopped =
       (std::accumulate(target_state.velocities.begin(), target_state.velocities.end(), 0.0) <= STOPPED_VELOCITY_EPS);
 
-  if (smoother_)
-  {
-    smoother_->reset(current_state.positions, current_state.velocities, current_state.accelerations);
```
Comment (sea-bass) [inline moveit_ros/moveit_servo/src/servo.cpp:639]: Not sure why `reset()` was being called here, but I think this was a bug?

File: moveit_ros/moveit_servo/src/servo_node.cpp (hunk 1 above)
Comment (marioprats) [reply moveit_ros/moveit_servo/src/servo_node.cpp:166]: Is `pauseServo` always called when switching to another mode? in other words, is it possible to move the arm with planners, while servo is active?

Comment (sea-bass) [reply moveit_ros/moveit_servo/src/servo_node.cpp:166]: Yes, this is possible because both the planning pipelines and servo are separate nodes and can be blasting out data to the same ros2_controller.

However, in most applications (MoveIt Studio very much included), we never have unpaused Servo and motion planning + execution happening at the same time.

I'd call this case "user error".

File: moveit_ros/moveit_servo/include/moveit_servo/servo.hpp [hunk 3]
```diff
@@ -125,7 +125,19 @@ class Servo
...
    * @return The next state stepping towards the required halting state.
    */
-  std::pair<bool, KinematicState> smoothHalt(const KinematicState& halt_state) const;
+  std::pair<bool, KinematicState> smoothHalt(const KinematicState& halt_state);
+
+  /**
+   * \brief Updates the smoother, if one exists, by filtering with a new state.
+   * @param state The state used to update the smoothing plugin.
+   */
+  void updateSmoother(KinematicState& state);
```
Comment (marioprats) [inline moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:134]: what about `doSmoothing` as in the smoother function it calls? or `applySmoothing`?
`updateSmoother` makes me think it's changing the underlying smoothing algorithm.

File: moveit_ros/moveit_servo/include/moveit_servo/servo.hpp [hunk 4]
```diff
@@ -125,7 +125,19 @@ class Servo
    * @param The last commanded joint states.
    * @return The next state stepping towards the required halting state.
    */
-  std::pair<bool, KinematicState> smoothHalt(const KinematicState& halt_state) const;
+  std::pair<bool, KinematicState> smoothHalt(const KinematicState& halt_state);
+
+  /**
+   * \brief Updates the smoother, if one exists, by filtering with a new state.
```
Comment (marioprats) [inline moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:131]: Maybe reword this to emphasize that this is applying smoothing to `state` and updating it accordingly.

**Review by marioprats — APPROVED** [review]
LGTM with minor comments

File: moveit_ros/moveit_servo/include/moveit_servo/servo.hpp [hunk 5]
```diff
@@ -215,6 +227,7 @@ class Servo
   std::atomic<double> collision_velocity_scale_ = 1.0;
   std::unique_ptr<CollisionMonitor> collision_monitor_;
 
+  // Pointer to the (optional) smoothing plugin.
```
Comment (AndyZe) [inline moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:230]: You could probably use std::optional here. Different issue, I suppose

**Review by AndyZe — CHANGES_REQUESTED** [review]
> this is the actual fix, the rest is reducing code duplication

I don't think this is true, LOC went up by 22 on a 2-line bug fix. Suggest reverting the new functions `doSmoothing()` and `resetSmoothing()` b/c they just add an extra layer. I think the smoother API is pretty clear already

**Comment by sea-bass** [comment]
> > this is the actual fix, the rest is reducing code duplication
> 
> I don't think this is true, LOC went up by 22 on a 2-line bug fix. Suggest reverting the new functions `doSmoothing()` and `resetSmoothing()` b/c they just add an extra layer. I think the smoother API is pretty clear already

Going by raw LOC, sure, more lines were added. However, the 2 line bug fix would have been 5 lines without this and would have copy pasted a bunch of code that already existed elsewhere:

```cpp
const auto current_state = getCurrentRobotState();
if (smoother_)
{
  smoother_->reset(current_state.positions, current_state.velocities, current_state.accelerations);
}
```

Also, I needed a public function so that the servo node could reset the smoother... meaning getting rid of at least the `resetSmoothing` function would require me to change `smoother_` to be a public member. Which is not ideal.

**Comment by AndyZe** [comment]
OK. I didn't realize this would involve making the member var public

**Review by AndyZe — APPROVED** [review]
